zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -0,0 +1,436 @@
+            is_valid = false;
+            continue;
+        };
+
+        // Check the index and root if the sapling note commitment tree at this height is a complete subtree.
+        if let Some((index, node)) = tree.completed_subtree_index_and_root() {
+            if subtree.index != index {
+                warn!("completed subtree indexes should match");
+                is_valid = false;
+            }
+
+            if subtree.node != node {
+                warn!("completed subtree roots should match");
+                is_valid = false;
+            }
+        }
+        // Check that the final note has a greater subtree index if it didn't complete a subtree.
+        else {
+            let Some(prev_tree) = db.sapling_tree_by_height(&subtree.end.previous()) else {
+                warn!(?subtree.end, "missing note commitment tree at subtree completion height");
+                is_valid = false;
+                continue;
+            };
+
+            let prev_subtree_index = prev_tree.subtree_index();
+            let subtree_index = tree.subtree_index();
+            if subtree_index <= prev_subtree_index {
+                warn!(
+                    ?subtree_index,
+                    ?prev_subtree_index,
+                    "note commitment tree at end height should have incremented subtree index"
+                );
+                is_valid = false;
+            }
+        }
+    }
+
+    let mut subtree_count = 0;
+    for (index, height, tree) in db
+        .sapling_tree_by_height_range(..)
+        .filter_map(|(height, tree)| Some((tree.subtree_index()?, height, tree)))
+        .filter_map(|(subtree_index, height, tree)| {
+            if tree.is_complete_subtree() || subtree_index.0 == subtree_count + 1 {
+                subtree_count += 1;
+                Some((subtree_index, height, tree))
+            } else {
+                None
+            }
+        })
+    {
+        let Some(subtree) = db.sapling_subtree_by_index(index) else {
+            warn!(?index, "missing subtree");
+            is_valid = false;
+            continue;
+        };
+
+        if subtree.index != index {
+            warn!("completed subtree indexes should match");
+            is_valid = false;
+        }
+
+        if subtree.end != height {
+            warn!(?subtree.end, "bad subtree end height");
+            is_valid = false;
+        }
+
+        if let Some((_index, node)) = tree.completed_subtree_index_and_root() {
+            if subtree.node != node {
+                warn!("completed subtree roots should match");
+                is_valid = false;
+            }
+        }
+    }
+
+    if !is_valid {
+        warn!("missing or bad sapling subtrees");
+    }
+
+    is_valid
+}
+
+/// Check that Orchard note commitment subtrees were correctly added.
+///
+/// Returns false if there are missing or bad subtrees.
+fn check_orchard_subtrees(db: &ZebraDb) -> bool {
```

Comment:
This is equivalent to `check_sapling_subtrees` except all instances of "sapling" have been replaced with "orchard".